File: source/Framework/Timer/TimerSignalObjectManager.cpp
/***************************************************************************
 *   Copyright (C) 2005 Vidyasagara Guntaka                                *
 *   All rights reserved.                                                  *
 *   Author : Amr Sabaa                                                    *
 ***************************************************************************/

#include "Framework/Timer/TimerSignalObjectManager.h"
#include "Framework/Timer/TimerWorker.h"
#include "Framework/Utils/AssertUtils.h"
#include <signal.h>
#include "Framework/ObjectModel/WaveAsynchronousContextForBootPhases.h"

namespace WaveNs
{

TimerSignalObjectManager::TimerSignalObjectManager ()
    : WaveLocalObjectManagerForUserSpecificTasks ("Framework Timer Signal")
{
}

TimerSignalObjectManager *TimerSignalObjectManager::getInstance ()
{
    static TimerSignalObjectManager *pTimerSignalObjectManager = NULL;

    if (NULL == pTimerSignalObjectManager)
    {
        pTimerSignalObjectManager = new TimerSignalObjectManager ();
        WaveNs::prismAssert (NULL != pTimerSignalObjectManager, __FILE__, __LINE__);
    }

    return (pTimerSignalObjectManager);
}

TimerSignalObjectManager::~TimerSignalObjectManager ()
{

}

PrismServiceId TimerSignalObjectManager::getPrismServiceId ()
{
    return ((getInstance ())->getServiceId ());
}

void TimerSignalObjectManager::catchTimerAlarm ()
{
    TimerWorker::processTimeOut ();
}

void TimerSignalObjectManager::boot (WaveAsynchronousContextForBootPhases *pWaveAsynchronousContextForBootPhases)
{
    trace (TRACE_LEVEL_DEVEL, "TimerSignalObjectManager::boot : Entering ...");

    pWaveAsynchronousContextForBootPhases->setCompletionStatus (WAVE_MESSAGE_SUCCESS);
    pWaveAsynchronousContextForBootPhases->callback ();

	prismCreateWaitableTimer ();

    while (1)
    {
        prismWaitForAlarmSignal ();
        catchTimerAlarm ();
    }
}

}

File: Java/source/com/CxWave/Wave/Framework/Timer/TimerSignalObjectManager.java
/***********************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 ***********************************************************************************************************/

package com.CxWave.Wave.Framework.Timer;

import com.CxWave.Wave.Framework.ObjectModel.WaveLocalObjectManager;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.NativeService;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.ObjectManagerPriority;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.PrePhaseService;
import com.CxWave.Wave.Framework.ObjectModel.Boot.WaveAsynchronousContextForBootPhases;
import com.CxWave.Wave.Framework.Type.WaveServiceId;
import com.CxWave.Wave.Framework.Utils.Assert.WaveAssertUtils;
import com.CxWave.Wave.Resources.ResourceEnums.ResourceId;
import com.CxWave.Wave.Resources.ResourceEnums.TraceLevel;
import com.CxWave.Wave.Resources.ResourceEnums.WaveObjectManagerPriority;

@PrePhaseService
@NativeService
@ObjectManagerPriority (WaveObjectManagerPriority.WAVE_OBJECT_MANAGER_PRIORITY_TIMER_SIGNAL)
public class TimerSignalObjectManager extends WaveLocalObjectManager
{
    private static TimerSignalObjectManager s_timerSignalObjectManager = null;

    private TimerSignalObjectManager ()
    {
        super (getServiceName ());
    }

    public static String getServiceName ()
    {
        return ("Framework Timer Signal");
    }

    public static TimerSignalObjectManager getInstance ()
    {
        if (null == s_timerSignalObjectManager)
        {
            s_timerSignalObjectManager = new TimerSignalObjectManager ();
        }

        WaveAssertUtils.waveAssert (null != s_timerSignalObjectManager);

        return (s_timerSignalObjectManager);
    }

    public static WaveServiceId getWaveServiceId ()
    {
        return ((getInstance ()).getServiceId ());
    }

    @Override
    public void boot (final WaveAsynchronousContextForBootPhases waveAsynchronousContextForBootPhases)
    {
        trace (TraceLevel.TRACE_LEVEL_DEVEL, "TimerSignalObjectManager::boot : Entering ...");

        waveAsynchronousContextForBootPhases.setCompletionStatus (ResourceId.WAVE_MESSAGE_SUCCESS);
        waveAsynchronousContextForBootPhases.callback ();
    }
}
